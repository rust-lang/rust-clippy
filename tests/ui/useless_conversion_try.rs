#![deny(clippy::useless_conversion)]

fn test_generic<T: Copy>(val: T) -> T {
    let _ = T::try_from(val).unwrap();
    val.try_into().unwrap()
}

fn test_generic2<T: Copy + Into<i32> + Into<U>, U: From<T>>(val: T) {
    // ok
    let _: i32 = val.try_into().unwrap();
    let _: U = val.try_into().unwrap();
    let _ = U::try_from(val).unwrap();
}

fn main() {
    test_generic(10i32);
    test_generic2::<i32, i32>(10i32);

    let _: String = "foo".try_into().unwrap();
    let _: String = TryFrom::try_from("foo").unwrap();
    let _ = String::try_from("foo").unwrap();
    #[allow(clippy::useless_conversion)]
    {
        let _ = String::try_from("foo").unwrap();
        let _: String = "foo".try_into().unwrap();
    }
    let _: String = "foo".to_string().try_into().unwrap();
    let _: String = TryFrom::try_from("foo".to_string()).unwrap();
    let _ = String::try_from("foo".to_string()).unwrap();
    let _ = String::try_from(format!("A: {:04}", 123)).unwrap();
    let _: String = format!("Hello {}", "world").try_into().unwrap();
    let _: String = "".to_owned().try_into().unwrap();
    let _: String = match String::from("_").try_into() {
        Ok(a) => a,
        Err(_) => "".into(),
    };
    // FIXME this is a false negative
    #[allow(clippy::cmp_owned)]
    if String::from("a") == TryInto::<String>::try_into(String::from("a")).unwrap() {}
}
