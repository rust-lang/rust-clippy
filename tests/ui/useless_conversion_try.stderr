error: useless conversion to the same type: `T`
  --> $DIR/useless_conversion_try.rs:4:13
   |
LL |     let _ = T::try_from(val).unwrap();
   |             ^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> $DIR/useless_conversion_try.rs:1:9
   |
LL | #![deny(clippy::useless_conversion)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider removing `T::try_from()`

error: useless conversion to the same type: `T`
  --> $DIR/useless_conversion_try.rs:5:5
   |
LL |     val.try_into().unwrap()
   |     ^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:27:21
   |
LL |     let _: String = "foo".to_string().try_into().unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:28:21
   |
LL |     let _: String = TryFrom::try_from("foo".to_string()).unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `TryFrom::try_from()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:29:13
   |
LL |     let _ = String::try_from("foo".to_string()).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `String::try_from()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:30:13
   |
LL |     let _ = String::try_from(format!("A: {:04}", 123)).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `String::try_from()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:31:21
   |
LL |     let _: String = format!("Hello {}", "world").try_into().unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:32:21
   |
LL |     let _: String = "".to_owned().try_into().unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: useless conversion to the same type: `std::string::String`
  --> $DIR/useless_conversion_try.rs:33:27
   |
LL |     let _: String = match String::from("_").try_into() {
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: aborting due to 9 previous errors

